File: a14/backend/app/routers/projects.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List

from ..database import get_db
from .. import crud, schemas

router = APIRouter(prefix="/api/projects", tags=["projects"])


@router.get("", response_model=List[schemas.Project])
def read_projects(skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    projects = crud.get_projects(db, skip=skip, limit=limit)
    return projects


@router.get("/{project_id}", response_model=schemas.Project)
def read_project(project_id: int, db: Session = Depends(get_db)):
    db_project = crud.get_project(db, project_id=project_id)
    if db_project is None:
        raise HTTPException(status_code=404, detail="Project not found")
    return db_project


@router.post("", response_model=schemas.Project)
def create_project(project: schemas.ProjectCreate, db: Session = Depends(get_db)):
    return crud.create_project(db=db, project=project)


@router.put("/{project_id}", response_model=schemas.Project)
def update_project(
    project_id: int, 
    project_data: schemas.ProjectUpdate, 
    db: Session = Depends(get_db)
):
    db_project = crud.update_project(db, project_id, project_data)
    if db_project is None:
        raise HTTPException(status_code=404, detail="Project not found")
    return db_project


@router.delete("/{project_id}")
def delete_project(project_id: int, db: Session = Depends(get_db)):
    success = crud.delete_project(db, project_id)
    if not success:
        raise HTTPException(status_code=404, detail="Project not found")
    return {"message": "Project deleted successfully"}


@router.get("/{project_id}/files", response_model=List[schemas.TrajectoryFile])
def get_project_files(project_id: int, db: Session = Depends(get_db)):
    db_project = crud.get_project(db, project_id=project_id)
    if db_project is None:
        raise HTTPException(status_code=404, detail="Project not found")
    return crud.get_project_files(db, project_id=project_id)


@router.get("/{project_id}/analysis", response_model=List[schemas.AnalysisResult])
def get_project_analysis(project_id: int, db: Session = Depends(get_db)):
    db_project = crud.get_project(db, project_id=project_id)
    if db_project is None:
        raise HTTPException(status_code=404, detail="Project not found")
    return crud.get_project_analysis(db, project_id=project_id)

File: a19/frontend/src/api/projects.ts
import { apiClient } from './client';
import { Project, Branch, CreateProjectData } from '@/types';

export const projectsApi = {
  async getProjects(): Promise<Project[]> {
    const response = await apiClient.get<{ projects: Project[] }>('/projects');
    return response.data.projects;
  },

  async getProject(id: string): Promise<Project> {
    const response = await apiClient.get<{ project: Project }>(`/projects/${id}`);
    return response.data.project;
  },

  async createProject(data: CreateProjectData): Promise<Project> {
    const response = await apiClient.post<{ project: Project }>('/projects', data);
    return response.data.project;
  },

  async updateProject(id: string, data: Partial<Project>): Promise<Project> {
    const response = await apiClient.put<{ project: Project }>(`/projects/${id}`, data);
    return response.data.project;
  },

  async deleteProject(id: string): Promise<void> {
    await apiClient.delete(`/projects/${id}`);
  },

  async getBranches(projectId: string): Promise<Branch[]> {
    const response = await apiClient.get<{ branches: Branch[] }>(`/projects/${projectId}/branches`);
    return response.data.branches;
  },

  async createBranch(projectId: string, name: string, baseBranch?: string): Promise<Branch> {
    const response = await apiClient.post<{ branch: Branch }>(`/projects/${projectId}/branches`, {
      name,
      baseBranch
    });
    return response.data.branch;
  }
};
